workspace/__lib__/xod/datetime/unpack-datetime/patch.cpp

```diff
@@ -0,0 +1,60 @@
+
+struct State {
+};
+
+// clang-format off
+{{ GENERATED_CODE }}
+// clang-format on
+
+bool isLeapYear(int y) {
+  return (1970 + y > 0) && !((1970 + y) % 4) && (((1970 + y) % 100) || !((1970 + y) % 400));
+}
+
+void evaluate(Context ctx) {
+    auto time = getValue<input_IN>(ctx);
+
+    static const uint8_t monthDays[] = { 31, 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31 };
+
+    emitValue<output_SEC>(ctx, time % 60);
+    time /= 60;
+    emitValue<output_MIN>(ctx, time % 60);
+    time /= 60;
+    emitValue<output_HOUR>(ctx, time % 24);
+    time /= 24;
+    emitValue<output_WDAY>(ctx, ((time + 4) % 7));
+
+    uint8_t year = 0;
+    uint32_t days = 0;
+
+    while ((unsigned)(days += (isLeapYear(year) ? 366 : 365)) <= time) {
+        year++;
+    }
+
+    emitValue<output_YEAR>(ctx, year + 1970);
+
+    days -= isLeapYear(year) ? 366 : 365;
+    time -= days;
+
+    uint8_t month = 0, monthLength = 0;
+    days = 0;
+
+    for (month = 0; month < 12; month++) {
+        if (month == 1) { // february
+            if (isLeapYear(year)) {
+                monthLength = 29;
+            } else {
+                monthLength = 28;
+            }
```

Comment:
Simpler expressed with:

```cpp
monthLength = isLeapYear(year) ? 29 : 28;
```

…and you’re calling `isLeapYear` on each iteration which is suboptimal. Suggest computing it once the `year` value settles down.

:open_mouth: And you’re using `uint8_t` to hold `year`. Does it even work while handling 2000-ths?